ecs-agent/acs/session/payload_responder.go

```diff
@@ -0,0 +1,162 @@
+
+package session
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+)
+
+const (
+	PayloadMessageName = "PayloadMessage"
+)
+
+type TaskMessageHandler interface {
+	ProcessTaskMessage(message *ecsacs.PayloadMessage,
+		ackFunc func(*ecsacs.AckRequest, []*ecsacs.IAMRoleCredentialsAckRequest)) error
+}
+
+// payloadResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.PayloadMessage messages sent by ACS.
+type payloadResponder struct {
+	taskMessageHandler TaskMessageHandler
+	respond            wsclient.RespondFunc
+}
+
+// NewPayloadResponder returns an instance of the payloadResponder struct.
+func NewPayloadResponder(taskMessageHandler TaskMessageHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &payloadResponder{
+		taskMessageHandler: taskMessageHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*payloadResponder) Name() string { return "payload responder" }
+
+func (r *payloadResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handlePayloadMessage
+}
+
+func (r *payloadResponder) handlePayloadMessage(message *ecsacs.PayloadMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", PayloadMessageName))
+
+	// Validate fields in the message.
+	if err := validatePayloadMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", PayloadMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle payload message.
+	err := r.taskMessageHandler.ProcessTaskMessage(message, r.ackFunc)
+	if err != nil {
+		logger.Critical(fmt.Sprintf("Unable to handle %s", PayloadMessageName), logger.Fields{
+			field.MessageID: aws.StringValue(message.MessageId),
+			field.Error:     err,
+		})
+	}
+}
+
+// ackFunc sends ACKs of the payload message and of the credentials associated with the tasks contained in the payload
+// message.
+func (r *payloadResponder) ackFunc(payloadAck *ecsacs.AckRequest, credsAcks []*ecsacs.IAMRoleCredentialsAckRequest) {
+	go r.sendAck(payloadAck)
+
+	for _, credsAck := range credsAcks {
+		go r.sendAck(credsAck)
+	}
+}
+
+// sendAck handles the sending of an individual specific ACK, assuming it is of type
+// ecsacs.IAMRoleCredentialsAckRequest or ecsacs.AckRequest.
+func (r *payloadResponder) sendAck(ackRequest interface{}) {
+	var credentialsAck *ecsacs.IAMRoleCredentialsAckRequest
+	var payloadMessageAck *ecsacs.AckRequest
+	credentialsAck, ok := ackRequest.(*ecsacs.IAMRoleCredentialsAckRequest)
+	if ok {
+		logger.Debug(fmt.Sprintf("ACKing credentials associated with %s", PayloadMessageName), logger.Fields{
+			"credentialsID": aws.StringValue(credentialsAck.CredentialsId),
```

Comment:
Nit: Maybe create a logger field as it is being used in multiple places.

Reply:
ACK, I'll add this in the next revision if any blocking changes need to be made. Thanks!

Reply:
Sounds good. Maybe add it in TODOs for the PR for the future reference. Thanks.